pkg/engine/armtype.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package engine
+
+import (
+	"encoding/json"
+
+	"github.com/Azure/azure-sdk-for-go/services/cosmos-db/mgmt/2015-04-08/documentdb"
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/mgmt/2018-02-14/keyvault"
+	"github.com/Azure/azure-sdk-for-go/services/preview/authorization/mgmt/2018-09-01-preview/authorization"
+	"github.com/Azure/azure-sdk-for-go/services/preview/msi/mgmt/2015-08-31-preview/msi"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2018-08-01/network"
+	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2018-02-01/storage"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-10-01/compute"
+)
+
+type ARMResource struct {
+	APIVersion string            `json:"apiVersion,omitempty"`
+	Copy       map[string]string `json:"copy,omitempty"`
+	DependsOn  []string          `json:"dependsOn,omitempty"`
+}
+
+type ARMResourceLocType struct {
+	ARMResource
+	Location string `json:"location,omitempty"`
+	Type     string `json:"type,omitempty"`
+}
+
+func (arm ARMResourceLocType) MarshalJSON() ([]byte, error) {
+	return json.Marshal(arm)
+}
+
+func (arm ARMResource) MarshalJSON() ([]byte, error) {
+	return json.Marshal(arm)
+}
+
+type VirtualMachineARM struct {
```

Comment:
this might be a dumb question but why the `ARM` name suffix on all these resources? isn't the fact that the file name is armtype sufficient? It seems redundant, is it following a specific naming convention? If not, slight preference for removing it.

Reply:
What you said definitely makes sense. It is what idiomatic go suggests. All of this should be in a separate package. But moving this to a separate package has other implications which I can't foresee completely. Open to changing this, if others have strong opinions on this. One condition though: We brainstorm :)

Reply:
I like carrying the extra overhead of delivering this semantic context in the name itself. Will guarantee readability of future code, so we are always crystal clear we're dealing with "ARM-wrapped representations of <insert resource type here>"